cluster/service_registry.go

```diff
@@ -45,10 +47,33 @@ func (sr *ServiceRegistry) Register(ctx context.Context, serviceName, nodeName,
 	}
 
 	key := filepath.Join(servicesPrefix, serviceName, nodeName)
-	if _, err = sr.KV.Put(ctx, key, string(val)); err != nil {
+
+	// TODO: Discuss appripriate lease time, I chose 2 seconds using no imperical data
+	resp, err := sr.cli.Grant(ctx, 2)
+	if err != nil {
+		return fmt.Errorf("failed to create lease for service: %w", err)
+	}
+
+	if _, err = sr.KV.Put(ctx, key, string(val), clientv3.WithLease(resp.ID)); err != nil {
 		return fmt.Errorf("failed to register node: %w", err)
 	}
 
+	keepAliveResp, err := sr.cli.KeepAlive(ctx, resp.ID)
+	if err != nil {
+		return fmt.Errorf("failed to keep service registered: %w", err)
+	}
+	go func() {
+		for {
+			kar := <-keepAliveResp
```

Comment:
I'd be more confident about preventing a goroutine leak if we also check for a cancelled context here.

```go
for {
  select {
   case kar := <-keepAliveResp: 
     // code
   case <-ctx.Done():
     return 
  }
}

Reply:
If the context gets cancelled the response will be an error. The reason I didn't want to exit when the context gets cancelled is because the keepAlive might not immeditly exit its loop in which case we still need to print its output.

I guess im only currently printing the ttl but maybe I should just print the whole response?

Reply:
I guess if the goal of this goroutine is to log the ttls then it can remain the same. Does it close when kar is `nil` or does the response have an error method/field?

Reply:
the response can have errors